Iter1

Fairness2 opened the pull request:
Грубая реализация без тестов

Fairness2 committed:
initial | Добавлены функции регистрации и авторизации пользователей.

В проект добавлены модели пользователей и обработчики для регистрации и аутентификации, а также логика работы с базой данных для хранения и извлечения пользовательских данных. Реализована структура директорий, включающая модели, хранилища, полезные нагрузки и обработчики ошибок. Были добавлены Dockerfile и конфигурации docker-compose для создания среды разработки и развертывания.

Измененные файлы:
- internal/models/user.go
- internal/payloads/common.go
- internal/payloads/register.go
- internal/repositories/usersql.go
- internal/repositories/user.go
- internal/repositories/executor.go
- internal/services/user/handlers.go
- cmd/server/main.go
- internal/helpers/helpers.go
- internal/helpers/helpers_test.go
- Dockerfile
- internal/commonerrors/requesterrors.go
- internal/database/migrations/sql/20241205174805_create_t_pass_tables_and_comment_table.sql
- docker-compose.yml
- internal/config/keys/public.pem
- internal/config/keys/private.pem

Fairness2 committed:
iter1.1 | Добавлены функции для дешифрования данных, начальный клиент.

Fairness2 committed:
iter1.2 | Реализованы новые функции работы с паролями, изменения в клиентах.

Созданы функции для управления паролями, включая дешифрование и обработку их отображения в клиенте. Внесены изменения в клиентскую часть для поддержки этих функций, добавлены новые структуры и исправлена логика использования токенов. Обновлены зависимости в проекте.

Список измененных файлов:
- internal/models/passwords.go
- internal/client/service/password.go
- cmd/client/main.go
- go.mod
- internal/services/content/password.go
- internal/client/models/password/list.go
- go.sum
- internal/client/serverclient/client.go
- internal/payloads/passwords.go
- internal/client/models/login/login.go
- internal/client/models/menu/menu.go

Fairness2 committed:
iter1.3 | Добавлены новые компоненты пользовательского интерфейса и встроенное шифрование паролей.

Созданы новые интерфейсные компоненты (Checkbox, Text Input, Text Area), добавлены методы для шифрования и дешифрования паролей в PasswordService, улучшена логика работы со списками, формами входа и управления паролями. Реализована интеграция шифрования и дешифрования на стороне клиента.

Измененные файлы:
- internal/client/components/inputs.go
- internal/client/service/login.go
- internal/client/components/checkbox.go
- internal/client/models/focusindex.go
- internal/client/service/password.go
- internal/client/user/user.go
- internal/client/models/password/list.go
- internal/encrypt/cipher/cipher.go
- internal/client/models/login/login.go
- internal/client/models/password/form.go

Fairness2 committed:
iter1.4 | Добавил функционал для работы с произвольными текстами.

Реализованы CRUD методы для управления произвольными текстами, включая подключение их к системе комментариев. Добавлены миграции для новой таблицы t_text, новые модели, сервисы и репозитории. Также обновлен текущий функционал для управления паролями, добавлены методы удаления с помощью нового подхода.

Измененные файлы:
- internal/payloads/text.go
- internal/models/comment.go
- internal/client/service/password.go
- internal/services/content/password.go
- internal/client/models/password/list.go
- internal/database/migrations/sql/20241215151031_create_t_text_table.sql
- internal/repositories/text.go
- internal/repositories/password.go
- internal/models/text.go
- internal/services/content/text.go

Fairness2 committed:
iter1.5 | Исправил баги и добавил поддержку текстов в клиентском приложении.

Исправлены ошибки в структуре и базовых методах репозиториев паролей и текстов. Добавлена полноценная поддержка работы с текстами в клиентской части, включая формы и списки, а также обновлен выбор в меню. Улучшена работа с шифрованием текстов для хранения в базе.

Измененные файлы:
- internal/services/content/password.go
- internal/services/content/text.go
- internal/repositories/password.go
- internal/repositories/text.go
- internal/client/models/text/form.go
- internal/client/models/text/list.go
- internal/client/service/text.go
- internal/client/models/menu/menu.go

Fairness2 committed:
iter1.6 | Добавил поддержку работы с данными банковских карт.

Реализована модель хранения данных банковских карт, включающая комментарии, и методы работы с ними в репозитории. Добавлены эндпоинты: создание, обновление, получение списка и удаление карт. Также внедрены миграции для таблицы карт в базе данных и маршруты для обработки запросов.

Измененные файлы:
- internal/models/comment.go
- internal/models/cards.go
- internal/payloads/cards.go
- internal/repositories/card.go
- internal/services/content/card.go
- internal/database/migrations/sql/20241216193119_create_t_card_table.sql
- internal/router/router.go

Fairness2 committed:
iter1.7 | Добавил интерфейсы для работы с банковскими картами в клиентской части.

Реализованы модели, сервисы и представления для создания, обновления, отображения списка и удаления карт в клиентском приложении. Обновлен главный экран для интеграции работы с картами.

Измененные файлы:
- internal/client/models/card/form.go
- internal/client/models/card/list.go
- internal/client/service/card.go
- internal/client/service/text.go
- internal/client/models/menu/menu.go

Fairness2 committed:
iter1.8 | Добавлена поддержка загрузки и управления файлами

Реализовано создание, обновление, удаление и управление файлами с комментариями. Добавлены новые таблицы, репозитории, модели, сервисы и обработчики для работы с файлами, включая работу с файлами на диске и их привязку к пользователям.

Измененные файлы:
- internal/database/migrations/sql/20241205174805_create_t_pass_tables_and_comment_table.sql
- internal/database/migrations/sql/20241217190010_create_t_file_table.sql
- internal/models/file.go
- internal/repositories/file.go
- internal/services/content/file.go
- internal/repositories/password.go
- internal/payloads/file.go
- internal/repositories/crud.go
- internal/client/service/file.go

Fairness2 committed:
iter1.9 | Добавлена поддержка работы с файлами

Реализовано создание, обновление, удаление, загрузка и просмотр файлов с комментариями через новые модели, формы и списки. Добавлены маршруты API, обновлены миграции базы данных, а также реализованы методы шифрования/дешифрования файлов.

Измененные файлы:
- internal/database/migrations/sql/20241217190010_create_t_file_table.sql
- internal/client/models/file/form.go
- internal/client/models/file/list.go
- internal/client/models/menu/menu.go
- internal/client/service/file.go
- internal/router/router.go
- internal/services/content/file.go
- go.mod
- go.sum
- .gitignore

Fairness2 committed:
iter1.10 | Удалена неиспользуемая конфигурация CLI, добавлена конфигурация времени сборки и поддержка пути загрузки.

Удалён устаревший механизм CLI конфигурации в клиентском коде, его заменили на использование параметров, определяемых на этапе сборки. Добавлен новый параметр пути для загрузки файлов (UploadPath) и возможность отображения информации о сборке в клиентской части.

Измененные файлы:
- internal/client/config/parse.go
- internal/config/parse.go
- cmd/client/main.go
- internal/config/config.go
- internal/client/models/login/login.go
- internal/client/config/config.go
- internal/router/router.go

Fairness2 committed:
iter1.11 | Удалён функционал шифрования и связанные с ним ключи и тесты.

Был удалён устаревший и больше не используемый функционал для шифрования данных, включая использование ключей и тесты для этого функционала. Эти изменения упрощают кодовую базу и устраняют ненужные зависимости.

Изменённые файлы:
- internal/config/keys/encrypt/private.pem
- internal/config/keys/encrypt/public.pem
- internal/encrypt/decrypter.go
- internal/encrypt/decrypter_test.go

Fairness2 committed:
iter1.12 | Реализована CRUD-сервисная структура для управления данными.

CRUD-функционал был реорганизован с внедрением обобщённого сервиса управления элементами, а также удалены и переработаны старые реализации. Это позволило сократить количество повторяющегося кода и унифицировать работу с различными типами данных.

Изменённые файлы:
- internal/client/models/password/list.go
- internal/client/service/text.go
- internal/payloads/passwords.go
- internal/client/service/card.go
- internal/payloads/file.go
- internal/client/service/file.go
- internal/client/models/file/form.go
- internal/client/models/file/list.go
- internal/client/models/card/form.go
- internal/payloads/cards.go
- internal/client/service/password.go

Fairness2 committed:
iter1.13 | Добавлены комментарии к коду и улучшена структура компонентов и сервисов.

К существующему коду добавлено множество описательных комментариев, уточняющих назначение методов, структур и их функциональности. Также были улучшены и унифицированы компоненты, CRUD-сервисы и модели, а часть функций дополнена новой логикой для повышения удобства работы.

Изменённые файлы:
- internal/client/components/inputs.go
- internal/client/service/login.go
- internal/client/components/checkbox.go
- internal/client/models/focusindex.go
- internal/client/models/password/list.go
- internal/encrypt/cipher/cipher.go
- internal/client/models/text/list.go
- internal/client/service/text.go
- internal/client/service/card.go
- internal/client/service/file.go
- internal/client/models/file/list.go
- internal/client/models/card/form.go
- internal/client/service/password.go
- cmd/client/main.go
- internal/config/config.go
- internal/client/serverclient/client.go

Fairness2 committed:
iter1.14 | Добавлена поддержка регистрации пользователей в сервисе и интерфейсе.

В сервисе авторизации добавлена логика для обработки регистрации пользователей с использованием флага isRegistration. В пользовательском интерфейсе добавлены кнопки для регистрации и обновлена навигация по элементам интерфейса, обеспечивая интуитивное поведение.

Изменённые файлы:
- internal/client/service/login.go
- internal/client/models/login/login.go

Fairness2 committed:
iter1.15 | Убран глобальный контекст из репозиториев и добавлены мок-тесты.

Из всех репозиториев удалено использование глобального контекста, заменив его явной передачей в методы. В сервисах произведена адаптация к изменениям репозиториев. Добавлены мок-тесты для интерфейсов, чтобы обеспечить тестирование функционала.

Изменённые файлы:
- internal/token/mocrep_test.go
- internal/services/content/password.go
- internal/repositories/file.go
- internal/services/content/file.go
- internal/repositories/executor.go
- internal/repositories/text.go
- internal/repositories/password.go
- internal/repositories/user.go
- internal/services/content/text.go
- internal/repositories/crud.go
- internal/repositories/card.go
- go.mod
- internal/token/middleware_test.go

Fairness2 committed:
iter1.16 | Добавлены новые моки и проведена рефакторинг методов.

В коде добавлены моки для интерфейсов, связанных с токенами и базой данных, что улучшает покрытие тестами. Проведен рефакторинг методов `Authenticator`, включая использование зависимостей через интерфейсы, что упрощает их тестирование и уменьшает связность кода.

Изменённые файлы:
- internal/token/mocrep_test.go
- internal/token/mocclaims_test.go
- go.mod
- internal/token/middleware_test.go
- internal/token/mocexecutor_test.go
- internal/token/middleware.go

ekubyshin commented on internal/client/models/file/form.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// InitialForm инициализирует и возвращает форму с предопределенными полями ввода и привязками помощи по навигации с помощью клавиатуры.
+func InitialForm(service *service.FileService, data *payloads.FileWithComment) Form {
```
предпочтительнее использовать DI на основе интерфейсов.

Fairness2 replied:
Сделал интерфейсы для внедрения объектов

ekubyshin commented on internal/client/models/login/login.go:
```diff
@@ -0,0 +1,169 @@
+func InitialModel(service *service.LoginService) Model {
+	m := Model{
+		inputs: []*components.TInput{
+			components.NewTInput("Логин", "", true),
```
При разработке интерфейсов все текстовые константы принято выносить в i18n файлы, чтобы потом билдить приложение с поддержкой мультиязычности. Это не является требованием, а просто в качестве информации.

ekubyshin commented on internal/config/parse.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+// NewConfig инициализирует новую консольную конфигурацию, обрабатывает аргументы командной строки
+func NewConfig() (*CliConfig, error) {
```
отличное разделение ответственности - конфиг и отдельно парсинг.

ekubyshin commented on internal/database/migrations/migrations.go:
```diff
@@ -0,0 +1,22 @@
+// Migrate применяет миграцию базы данных с использованием встроенных файлов миграции и диалекта PostgreSQL.
+// Принимает соединение с базой данных (*sql.DB) в качестве входных данных и возвращает ошибку, если миграция не удалась.
+func Migrate(db *sql.DB) error {
+	goose.SetBaseFS(embedMigrations)
```
подключены миграции - отлично

ekubyshin commented on internal/database/pool.go:
```diff
@@ -0,0 +1,79 @@
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxOpenConns(maxConnections)
```
не забыл настроить коннекшен пул - отлично!

ekubyshin commented on internal/encrypt/cipher/cipher.go:
```diff
@@ -0,0 +1,75 @@
+
+// Cipher — это структура, в которой хранится криптографический ключ для операций шифрования и дешифрования.
+type Cipher struct {
+	Key []byte
```
имеет ли смысл делать поле публичным?

Fairness2 replied:
Сделал не публичным

ekubyshin commented on internal/models/cards.go:
```diff
@@ -0,0 +1,29 @@
+// CardContent представляет собой запись информации о карте, хранящуюся в базе данных.
+// Он включает поля для сведений о карте и связанных с ними метаданных, таких как отметки времени создания и обновления.
+type CardContent struct {
+	ID        int64     `db:"id"`
```
поле и так публичное, поэтому не вижу смысла делать дополнительно методы геттеры, либо надо делать поля приватными и отдельно геттеры и сеттеры. В данной реализации не понятна суть геттера

Fairness2 replied:
Геттер нужен для удовлетворения интерфейсу для сохранения в базе данных

ekubyshin commented on internal/models/cards.go:
```diff
@@ -0,0 +1,29 @@
+// Он включает поля для сведений о карте и связанных с ними метаданных, таких как отметки времени создания и обновления.
+type CardContent struct {
+	ID        int64     `db:"id"`
+	UserID    int64     `db:"user_id"`
```
на мой взгляд в данном проекте для айдишек оправдано использовать UUID, чтобы проще было резолвить конфликты, а не заниматься инкрементом айдишек

Fairness2 replied:
Перевёл контентные модели на uuid-v4

ekubyshin commented on internal/payloads/passwords.go:
```diff
@@ -0,0 +1,54 @@
+		return err
+	}
+	item.Username = eUsername
+	ePass, err := ch.Encrypt(item.Password.Password)
```
шифровани осуществляется по ключу? Те на сервере хранится и данные и сам ключ, что несколько снижает безопасность самих данных. Есть ключи, которые не играют роли без мастер-пароля, что делает ключ бессмысленным без мастер-пароля, а следовательно его можно хранить где угодно.

Fairness2 replied:
Сейчас шифрование происходит по паролю пользователя. На сервер поступает уже зашифрованная информация, и при передаче на клиент рассшифровывается.

ekubyshin replied:
если будет время и желание, то предлагаю поизучать тему шифрования по ключу с паролем.

Fairness2 replied:
Сделал формирование ключа с помощью PBKDF2, вроде как он используется для формирования ключа из мастер пароля. 
https://thecode.media/kak-ustroeny-menedzhery-parolej-v-chyom-ih-osobennosti-polza-i-opasnosti/

ekubyshin commented on internal/router/router.go:
```diff
@@ -0,0 +1,95 @@
+
+// NewRouter конфигурация роутинга приложение
+func NewRouter(dbPool *database.DBPool, cnf *config.CliConfig) chi.Router {
+	lHandlers := user.NewHandlers(dbPool.DBx, cnf.JWTKeys, cnf.TokenExpiration, cnf.TokenExpiration, cnf.HashKey)
```
довольно раздутый набор параметров, мб имеет смысл сделать структуру с конфигой и использовать один из паттернов для работы: билдер или опшенал функции.

Fairness2 replied:
Сделал структуру настройки

ekubyshin commented on internal/serverapp/app.go:
```diff
@@ -0,0 +1,58 @@
+	// Вызываем функцию закрытия базы данных
+	defer pool.Close()
+	// Производим миграции базы
+	if err = pool.Migrate(); err != nil {
```
если получится, то применение миграций лучше вытащить в отдельный гитхаб экшен и не завязывать его на запуск приложения. Это позволит настроить процесс деплоя более гибко.
В реальных проектах миграции принято делать отдельным экшеном в пайплайне, чтобы при необходимости можно было их откатить.

Fairness2 replied:
Вынес миграции в гитхаб экшен

ekubyshin commented on internal/serverapp/app.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	wg := new(errgroup.Group)
+	serv := server.NewServer(ctx, router.NewRouter(pool, cnf), cnf.Address)
```
не хватает обработки signal.Notify

Fairness2 replied:
Используется контекст signal.NotifyContext, так что он отменяется по приходу сигнала

ekubyshin replied:
да выше не увидел

Fairness2 committed:
iter1.17 | Добавлены новые константы ошибок и модульные тесты для обработчиков.

Добавлены константы `ErrPasswordEmpty` и `ErrHashEmpty` для улучшения читаемости и централизованного определения ошибок. Реализованы новые модульные тесты для методов обработки пользователей, включая регистрацию, аутентификацию, генерацию токенов и проверки существования пользователя, что обеспечивает более полное покрытие тестами.

Изменённые файлы:
- internal/models/user.go
- internal/services/user/handlers_test.go

Fairness2 committed:
iter1.18 | Заменены зависимости репозитория на интерфейсы и добавлены макеты для тестирования.

Изменена структура сервисов и репозиториев: использованы интерфейсы вместо конкретных типов, что улучшает тестируемость. Добавлены мок-объекты для unit-тестов с использованием GoMock. Также произведен рефакторинг методов обработки HTTP-запросов для упрощения и переиспользования кода.

Измененные файлы:
- internal/services/content/moccardrepo_test.go
- internal/services/content/mocfilerepo_test.go
- internal/services/content/mocpasswordrepo_test.go
- internal/services/content/moctextrepo_test.go
- internal/services/content/password.go
- internal/services/content/file.go
- internal/services/content/text_test.go
- internal/repositories/mocexecutor_test.go
- internal/repositories/executor.go
- internal/repositories/text.go
- internal/repositories/password.go
- internal/repositories/file.go

Fairness2 committed:
iter1.20 | Добавлены тесты для репозиториев и настройки миграций.

Внедрены unit-тесты для репозиториев пользователей, включая методы проверки существования, создания, обновления и извлечения данных. Добавлен workflow GitHub Actions для выполнения миграций базы данных. Реализованы изменения в интерфейсах ITX и Mock для улучшения тестирования с использованием GoMock.

Измененные файлы:
- .github/workflows/migrations.yml
- internal/repositories/user_test.go
- internal/repositories/crud.go
- internal/token/mocexecutor_test.go

Fairness2 committed:
iter1.19 | Добавлены тесты для логгирования и сервисов управления карточками.

Внедрены unit-тесты для middleware логгирования и сервисов управления карточками (создание, обновление, удаление, получение). Использованы макеты репозиториев на основе GoMock для обеспечения изолированного, контролируемого тестирования. Это позволяет упростить проверку корректности логики и повысить надежность кода.

Измененные файлы:
- internal/logger/middleware_test.go
- internal/services/content/card_test.go

Fairness2 committed:
Merge remote-tracking branch 'origin/iter1' into iter1

Fairness2 committed:
iter1.21 | Добавлены тесты для различных репозиториев.

Созданы unit-тесты для репозиториев карточек, текстовых данных, паролей и файлов. Тесты покрывают создание, извлечение и удаление данных, а также обработку ошибок и работу с mock-объектами. Это существенно повышает надежность и тестируемость логики данных.

Измененные файлы:
- internal/repositories/card_test.go
- internal/repositories/text_test.go
- internal/repositories/password_test.go
- internal/repositories/file_test.go
- internal/repositories/crud_test.go

Fairness2 committed:
iter1.22 | Реализована поддержка интерфейсов шифрования и дешифрования, добавлены тесты.

Добавлены интерфейсы Encrypter и Decrypter для подключения шифровщиков. Переписаны методы Encrypt и Decrypt для работы с интерфейсами вместо конкретной реализации. Также созданы и интегрированы unit-тесты для тестирования текстов, паролей, файлов и карточек на основе mock-объектов, что повышает тестируемость и надежность кода.

Измененные файлы:
- internal/payloads/text.go
- internal/serverapp/app.go
- internal/payloads/passwords.go
- internal/payloads/file.go
- internal/payloads/encrypterdecrypter.go
- internal/payloads/mocdecrypter_test.go
- internal/payloads/text_test.go
- internal/database/pool.go
- internal/payloads/passwords_test.go
- internal/payloads/cards.go
- internal/payloads/file_test.go
- internal/payloads/cards_test.go

Fairness2 committed:
iter1.23 | 80% покрытия тестами без клиента. Переписаны биндинги окружения и обработка ошибок в конфигурации, добавлены тесты.

Оптимизирована функция bindEnv для использования MustBindEnv вместо BindEnv для сокращения кода и упрощения обработки ошибок. Добавлены новые ошибки ErrNoPrivateKey и ErrNoPublicKey для настройки ключей RSA, а также реализованы unit-тесты для повышения надежности конфигурационного модуля.

Измененные файлы:
- .github/workflows/migrations.yml
- internal/config/parse.go
- internal/config/parse_test.go

Fairness2 committed:
iter1.24 | Интеграция интерфейсов для сервисных модулей и рефакторинг методов.

Внедрены интерфейсы для унификации сервисных модулей (processService, loginService и др.), что позволило упростить взаимодействие между компонентами и снизить зависимость от конкретных реализаций. Переписаны методы, такие как NewCRUDService, с учетом новых интерфейсов, улучшена документация для API-методов и обработчиков. Также добавлены аннотации Swagger для текстового, файлового и паролей сервисов, что способствует лучшей интеграции и структурированности кода.

Измененные файлы:
- internal/client/service/login.go
- internal/services/content/password.go
- internal/client/models/password/list.go
- internal/client/models/text/list.go
- internal/client/service/text.go
- internal/services/content/file.go
- internal/client/service/card.go
- internal/services/content/text.go
- internal/client/service/file.go
- internal/client/models/file/form.go
- internal/client/models/file/list.go

Fairness2 committed:
iter1.25 | Удалил старую реализацию карт и добавил генерацию Swagger документации.

Удалён файл с реализацией моделей карт и шифрования (`internal/payloads/cards.go`), так как он больше не используется. Добавлен новый файл с автоматически сгенерированной Swagger документацией (`docs/docs.go`) для обеспечения актуальности API и упрощения его интеграции.

Измененные файлы:
- internal/payloads/cards.go (удален)
- docs/docs.go (добавлен)

Fairness2 committed:
iter1.26 | Исправил описание хэндлеров и убрал дублирующийся текст.

Обновлены описания методов в комментариях, исправлены ошибки и устранены дублирования в аннотациях для Swagger. Эти изменения обеспечивают более точную документацию API и упрощают понимание кода.

Измененные файлы:
- internal/services/user/handlers.go

ekubyshin commented on .github/workflows/migrations.yml:
миграции в отдельном экшене - отлично!

ekubyshin commented on internal/helpers/helpers.go:
```diff
@@ -0,0 +1,77 @@
+// - response: http.ResponseWriter object containing information about the HTTP response
+// - status: the HTTP status code to set in the response
+// - message: the message to write to the response
+func SetHTTPResponse(response http.ResponseWriter, status int, message []byte) error {
```
на самом деле складывать кучу всего в какие-то общие пакеты, например, utils, helpers и тд, не очень хорошая практика, а лучше создавать пакеты, которые будут отражать предназначение. Со временем, подобные пакеты разрастаются до монструозных размеров, тк все кому не лень, все что не относится к конкретной бизнес логике6 будут складывать туда все излишки ))

Fairness2 replied:
Так как в этом пакете содержаться только функции для отправки ответа, то переименовал пакет в соответствии с его назначением

ekubyshin commented on internal/middlewares/middlewares.go:
```diff
@@ -0,0 +1,14 @@
+	"net/http"
+)
+
+// JSONHeaders Устанавливаем заголовки свойственные методам с JSON
```
здесь тоже лучше не складывать в одном файле, а называть каждый файл отдельным мидлвар

Fairness2 replied:
Переименовал пакет, так как он содержит только один определённый мидлвар

ekubyshin commented on internal/payloads/encrypterdecrypter.go:
```diff
@@ -0,0 +1,11 @@
+package payloads
+
+// Encrypter — это интерфейс, определяющий метод шифрования фрагментов байтов и возврата зашифрованного результата или ошибки.
+type Encrypter interface {
```
выглядит продуманно

ekubyshin commented on internal/repositories/adapter.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+// DBAdapter Структура которая скрывает реализацию подключения к бд за функциями, которые можно замокировать
+type DBAdapter struct {
```
если очень хочется мокировать, то можно посмотреть на что-то готовое, например sqlmock. На мой взгляд, тестировать логику работы с бд на моках - не лучшее решение.
1. не позволит проверять код автотестами, вместе с бд
2. не понятно, насколько код будет работать на реальной бд
для подобных целей, я использую gotestcontainers

Fairness2 replied:
Продублировал часть тестов с помощью тест контейнеров с реальной базой

ekubyshin commented on internal/router/router.go:
```diff
@@ -0,0 +1,104 @@
+		RefreshTokenExpiration: cnf.TokenExpiration,
+		HashKey:                cnf.HashKey,
+	}
+	lHandlers := user.NewHandlers(userCnf)
```
отлично, что сделал отдельную абстракцию

ekubyshin commented on internal/router/router.go:
```diff
@@ -0,0 +1,104 @@
+	/*router.Get("/swagger/*", httpSwagger.Handler(
+		httpSwagger.URL("/swagger/doc.json"), //The url pointing to API definition
+	))*/
+	router.Route("/api/user", func(r chi.Router) {
```
можно сделать отдельные слои, чтобы каждый сервис возвращал свой subroute, который просто можно замаунтить здесь. Как вариант реализации

Fairness2 replied:
Перенёс инсталяцию роутов в объекты хэндлеров

ekubyshin commented on internal/services/content/card.go:
```diff
@@ -0,0 +1,234 @@
+		return
+	}
+	// Для создания нельзя передавать идентификатор
+	if body.ID != 0 {
```
существуют различные готовые пакеты, для реализации валидации входящих данных, чтобы не приходилось это делать в каждо хэндлере. Будет желание, можно поизучать этот вопрос

Fairness2 replied:
Использовал https://github.com/asaskevich/govalidator

ekubyshin commented on internal/services/content/card.go:
```diff
@@ -0,0 +1,234 @@
+		helpers.ProcessResponseWithStatus("User not found", http.StatusUnauthorized, response)
+		return
+	}
+	card := models.CardContent{
```
я бы еще добавил валидаторов на каждую модель

Fairness2 replied:
Добавил валидаторы на пэйлоады, кажется, что в них должны они быть

ekubyshin commented on internal/services/content/card.go:
```diff
@@ -0,0 +1,234 @@
+		helpers.ProcessResponseWithStatus("User not found", http.StatusUnauthorized, response)
+		return
+	}
+	cards, err := s.repository.GetByUserID(request.Context(), user.ID)
```
Если очень хочется и позволяет время, то можно попробовать реализовать кэш с использованием какого-нибудь паттерна, например, https://www.gomomento.com/blog/6-common-caching-design-patterns-to-execute-your-caching-strategy/ Но это не обязательно

ekubyshin commented on internal/services/content/password.go:
```diff
@@ -0,0 +1,249 @@
+
+// PasswordService предоставляет методы для управления паролями пользователей и соответствующими комментариями в системе.
+type PasswordService struct {
+	repository passwordRepository
```
отлично - интерфейсы!

ekubyshin commented on internal/services/content/password.go:
```diff
@@ -0,0 +1,249 @@
+// NewPasswordService инициализирует и возвращает новый экземпляр PasswordService, настроенный с использованием предоставленной базой.
+func NewPasswordService(dbPool repositories.SQLExecutor) *PasswordService {
+	return &PasswordService{
+		repository: repositories.NewPasswordRepository(dbPool),
```
на мой взгляд, лучше сразу зависить от интерфейса, а где то  в другом месте создавать зависимости и пробрасывать их в сервис через внедрение

Fairness2 replied:
Избавился от таких внедрений

Fairness2 committed:
iter1.27 | Реализовал использование безопасного шифрования и обновил структуру пользователя.

Была добавлена новая функция для настройки пользователя с использованием шифровальных механизмов. Рефакторинг включил защиту пароля через PBKDF2, модификацию тестов и внедрение новых конфигурационных параметров для шифрования, что повышает безопасность данных.

Измененные файлы:
- internal/client/service/login.go
- internal/encrypt/cipher/cipher_test.go
- internal/client/user/user.go
- internal/encrypt/cipher/cipher.go
- internal/client/service/crud.go
- internal/client/config/config.go
- internal/client/service/file.go

Fairness2 committed:
iter1.28 | Перевёл приложение на UUID. Добавил валидацию ввода пользователя. Избавился от сложных зависимостей. Перенёс регистрацию роутов в хэндлеры.

Fairness2 committed:
iter1.29 | Добавлены интеграционные тесты с использованием testcontainers. Расширено покрытие существующих тестов.

Добавлены тесты для взаимодействия с реальной базой данных через testcontainers. Обновлены существующие тесты для использования более гибкой структуры тестовых данных. Обновлены зависимости в `go.mod` для поддержки новых библиотек, необходимых для тестирования.

Измененные файлы:
- internal/repositories/common_test.go
- internal/repositories/user_test.go
- internal/repositories/crud_test.go
- go.mod

Fairness2 committed:
iter1.30 | Выключил миграции в приложении)

Fairness2 committed:
iter1.31 | Добавил README файл.

Обновлены комментарии в файле конфигурации для большей унификации и повышения читаемости. Скорректированы типы переменных в клиентской конфигурации для большей строгости и совместимости. README дополнен расширенным описанием проекта и его технической реализации.

Изменённые файлы:
- internal/config/config.go
- internal/client/config/config.go
- README.md

ekubyshin commented on internal/client/models/card/form.go:
```diff
@@ -0,0 +1,267 @@
+			return m, tea.Quit
+		// Set focus to next input
+		case "tab", "shift+tab", "enter", "up", "down":
+			s := msg.String()
```
тут можно улучшить и разбиить на отдельные методы, тк если логика будет больше, то получится довольно монструозный код

Fairness2 replied:
Разбил на методы

ekubyshin commented on internal/client/models/card/form.go:
```diff
@@ -0,0 +1,267 @@
+		m.modelError = err
+		return m, m.getCmds()
+	}
+	if m.data.ID == "" {
```
можно улучшить, воспользовавшись ошибкой err выше и в нее складывать результат вызова crteate/update, а обработку уже один раз ниже else

Fairness2 replied:
Поправил в формах

ekubyshin commented on internal/client/models/card/form.go:
```diff
@@ -0,0 +1,267 @@
+	}
+	fmt.Fprintf(&b,
+		`
+ %s
```
выглядит очень магически )) мб воспольользоваться шаблонами или как-то вынести в константы?

Fairness2 replied:
Вынес в константу шаблон

ekubyshin commented on internal/client/models/file/form.go:
```diff
@@ -0,0 +1,181 @@
+
+// InitialForm инициализирует и возвращает форму с предопределенными полями ввода и привязками помощи по навигации с помощью клавиатуры.
+func InitialForm(service processService, data *payloads.FileWithComment) Form {
+	fp := components.NewTInput("Путь к файлу", "", true)
```
все таки предлагаю вынести константы интерфейса в отдельный файл с константами, и где надо переиспользовать.

Fairness2 replied:
Вынес все надписи в константы в пакете, а те, что используются в нескольких, в общий пакет

ekubyshin commented on internal/client/models/file/list.go:
```diff
@@ -0,0 +1,346 @@
+
+// processService определяет интерфейс для управления файлами, включая шифрование, дешифрование, создание, обновление и удаление.
+// Он предоставляет методы для управления данными файла с соответствующими комментариями и поддерживает шифрование и дешифрование на уровне файла.
+type processService interface {
```
не очень нравится этот подход писать интерфейс в этом пакете, а не там где он используется. По го стилю, лучше кончено выносить там где используется и делать интерфейсы максимум 1-2 метода, а тут довольно раздутый интерфейс. Interface pollution. 
Можно ознакомиться в книге Go 101 Mistakes, раздел Interface Pollution

Fairness2 replied:
В каждом пакете был свой processService, но я разделил его на интерфейс для формы и для листа, так же в каждом пакете свой

ekubyshin commented on internal/client/models/file/list.go:
```diff
@@ -0,0 +1,346 @@
+		l.modelError = err
+		return l, nil
+	}
+	if err = l.pService.DecryptFile(temp, destFile); err != nil {
```
здесь может случиться уязвимость в безопасности, тк сохраняется дешифрованый файл, а к нему может получить доступ злоумышленни

Fairness2 replied:
Этот метод предназначен для получения пользователем его сохранённого файла, чтобы его отдать, нужно его расшифровать. Расшифрованный файл сохраняется в то расположение, который указывает пользователь или по умолчанию

Fairness2 committed:
iter1.32 | Рефакторинг моделей, сервисов и форм, добавлены шаблоны строк.

Обновлены структуры форм и списков для оптимизации UI-логики с использованием шаблонов строк. Введены новые сервисы для управления объектами по умолчанию, повышая модульность и переиспользуемость кода. Добавлены константы и унифицирован текстовый стиль для согласованности интерфейса.

Изменённые файлы:
- internal/client/models/focusindex.go
- internal/client/models/password/list.go
- internal/client/models/text/list.go
- internal/client/service/text.go
- internal/client/service/card.go
- internal/client/service/file.go
- internal/client/models/file/form.go
- internal/client/models/file/list.go

Fairness2 committed:
iter1.33 | Унифицированы типы и улучшена навигация в формах, добавлены комментарии.

Изменены типы с `interface{}` на `any` для унификации и повышения читаемости кода. Улучшена навигационная логика в формах, добавлены вспомогательные функции для обработки событий и обновления UI. Добавлены поясняющие комментарии для ключевых структур и методов в сервисах и моделях.

Изменённые файлы:
- internal/logger/middleware_test.go
- internal/repositories/crud_test.go
- internal/client/models/focusindex.go
- internal/client/models/file/form.go
- internal/client/models/file/list.go
- internal/client/models/password/list.go
- internal/client/models/text/list.go
- internal/client/service/text.go
- internal/client/service/card.go
- internal/client/service/file.go
- internal/logger/logger.go
- internal/repositories/executor.go
- internal/payloads/mocdecrypter_test.go
- internal/validators/validators.go
- internal/client/models/card/form.go
- internal/client/service/password.go
- internal/client/models/card/list.go
- internal/client/user/user.go
- internal/client/models/text/form.go
- internal/token/generator.go
- internal/client/models/login/login.go

Fairness2 committed:
iter1.34 | Удалены неиспользуемые переменные.

Изменённые файлы:
- internal/client/models/card/list.go

Fairness2 committed:
iter1.35 | Добавлена функциональность возврата к предыдущей модели.

Добавлен интерфейс Backable для возврата к последней модели через "esc", обновлены формы и списки элементов для поддержки данной функциональности. Удалены неиспользуемые зависимости и улучшена навигация ввода в формах.

Изменённые файлы:
- internal/client/models/file/form.go
- internal/client/models/file/list.go
- internal/client/models/card/form.go
- internal/client/models/card/list.go
- internal/client/models/focusindex.go
- internal/client/models/backable.go
- internal/client/models/text/form.go
- internal/client/models/text/list.go
- internal/client/models/password/form.go
- internal/client/models/password/list.go

Fairness2 committed:
iter1.36 | Добавлены комментарии к Backable

Изменённые файлы:
- internal/client/models/backable.go
